File: typescript/packages/bench/src/repositories/warrantyRepository.ts
/** Read side over warranty claims. */

import type { Repository } from "@atelier/core";

/** One warranty claim raised against an order. */
export interface WarrantyClaim {
    readonly id: number;
    readonly repairOrderId: number;
    readonly settled: boolean;
}

/** Read side over warranty claims. */
export class WarrantyRepository implements Repository<number, WarrantyClaim> {
    constructor(private readonly records: readonly WarrantyClaim[] = []) {}

    find(id: number): WarrantyClaim | undefined {
        return this.records.find((record) => record.id === id);
    }

    all(): readonly WarrantyClaim[] {
        return this.records;
    }

    count(): number {
        return this.records.length;
    }

    /** Claims still awaiting a decision. */
    pending(): readonly WarrantyClaim[] {
        return this.records.filter((claim) => !claim.settled);
    }
}

File: rust/crates/atelier-bench/src/repositories/warranty_repository.rs
//! Read side over warranty claims.

use atelier_core::contracts::repository::Repository;

/// One warranty claim raised against an order.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct WarrantyClaim {
    pub id: u32,
    pub repair_order_id: u32,
    pub settled: bool,
}

/// Read side over warranty claims.
#[derive(Debug, Clone, Default)]
pub struct WarrantyRepository {
    records: Vec<WarrantyClaim>,
}

impl WarrantyRepository {
    pub fn new(records: Vec<WarrantyClaim>) -> Self {
        Self { records }
    }

    /// Claims still awaiting a decision.
    pub fn pending(&self) -> Vec<&WarrantyClaim> {
        self.records.iter().filter(|claim| !claim.settled).collect()
    }
}

impl Repository for WarrantyRepository {
    type Id = u32;
    type Record = WarrantyClaim;

    fn find(&self, id: Self::Id) -> Option<&Self::Record> {
        self.records.iter().find(|record| record.id == id)
    }

    fn all(&self) -> &[Self::Record] {
        &self.records
    }
}
